python/oneflow/test/modules/test_shutting_down.py

```diff
@@ -0,0 +1,42 @@
+"""
+Copyright 2020 The OneFlow Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+import oneflow
+import os
+
+
+world_size = os.getenv("WORLD_SIZE")
+
+
+class TestCallWhenShuttingDown:
+    def __init__(self):
+        self.oneflow = oneflow
+        tensor = oneflow.ones((2, 2))
+        print(tensor)
+
+    def __del__(self):
+        if world_size == 1:
+            tensor = self.oneflow.ones((2, 2))
+
+
+test_call_when_shutting_down = TestCallWhenShuttingDown()
+
+
+class TestSyncWhenShuttingDown:
+    def __del__(self):
+        oneflow._oneflow_internal.eager.Sync()
```

Comment:
这里也有可能触发找不到 oneflow module的问题。


### 另外一种推荐写法
找到了，通过kwargs把module捕获到函数的locals中，好像更美观一点：

```
class TestSyncWhenShuttingDown:
    def __del__(self, of=oneflow):
        print("locals", locals().keys())
        print("of", of)
        of._oneflow_internal.eager.Sync()
```

### module析构的顺序不确定
`__del__`在atexit之后执行，module析构的顺序不确定，看起来是python的老大难问题，python自己都不做保证。

> [__del__()](https://docs.python.org/3.7/reference/datamodel.html#object.__del__) can be executed during interpreter shutdown. As a consequence, the global variables it needs to access (including other modules) may already have been deleted or set to None. 

> Python guarantees that globals whose name begins with a single underscore are deleted from their module before other globals are deleted;

这个技巧试了下，没用。


> Bugs and caveats: The destruction of modules and objects in modules is done in random order; this may cause destructors ([__del__()](https://docs.python.org/3.7/reference/datamodel.html#object.__del__) methods) to fail when they depend on other objects (even functions) or modules.

别的项目也遇到类似的问题，采用了放弃治疗的做法，发现是shutting down，就直接返回：https://github.com/cupy/cupy/pull/2809

参考资料：
- https://docs.python.org/3.7/reference/datamodel.html#object.__del__
- https://docs.python.org/3.7/c-api/init.html#c.Py_FinalizeEx

### python下的资源依赖和循环引用
另外python gc对于定义了del的对象，如果有循环引用的问题，就不做析构。
- 为此python也提供了weak ref：https://docs.python.org/3/library/weakref.html 
- 另外对于避免循环引用，建议采用“依赖注入”，就是后面PR，graph那里的引用计数来显式管理资源依赖的实现方式；